### ISSUE TITLE
Fedora QA cooperation

### ISSUE THREAD

**Issue by kparal** [coreos/fedora-coreos-tracker#17]
Note: I originally started this as a discussion at [Fedora discourse](https://discussion.fedoraproject.org/t/fedora-qa-cooperation/173), now moving it here after request.

----

Hello, my name is Kamil Páral and I'm part of the Fedora QA team. I was asked to be a point of contact between Fedora CoreOS and Fedora QA. I would like to learn what the current state of QA is in your project, and how I can help you with integrating it into Fedora QA processes, or extending it.

I know there hasn't been any Fedora CoreOS release yet, but do you have any QA processes existing already? Testcases to describe what to test and how, testplans to describe which tests to run when, automated tests executed regularly (before release? after each commit?), results publicly visible somewhere? You know, the usual boring QA stuff.

As an example, here's our Cloud-specific test plan containing just a few test cases that we used for Fedora 28 Cloud/Atomic release validation:
https://fedoraproject.org/wiki/Test_Results:Fedora_28_RC_1.1_Cloud
(scroll to the bottom). These were tested manually during release. Atomic folks had some automated testing as well, I believe, but I don't really know the details (I'm sure the right people in your team will know about it).

If it makes sense to you, we could adopt those test cases, extend them to cover the most important CoreOS functionality, and create a separate CoreOS section for it. If you have any automated test results, we can talk about how to best integrated into our usual workflows. For the parts that we have automated in our team, we usually combine writing test results directly to such wiki matrices as shown above, and also inspect the failures in the tools' specific frontend. For example, everything filled out by _coconut_ in our [Installation matrix](https://fedoraproject.org/wiki/Test_Results:Fedora_28_RC_1.1_Installation) has been automatically tested by [OpenQA](https://openqa.fedoraproject.org/) - the failures are examined in its frontend directly. But there are definitely other approaches that can be taken.

I can also help you set up [test days](https://fedoraproject.org/wiki/QA/Test_Days) for your project, and try to reach out to Fedora community to help you get more of a unique test coverage. Or.... tell me what else you'd like to see or help with.

Looking forward to your feedback!

**Comment by arithx** [comment]
Our current plan is to utilize the [`kola` test framework](https://github.com/coreos/mantle) for testing Fedora CoreOS, which was originally designed for testing clusters of Container Linux machines. It encompasses everything from the machine creation on different platforms, the actual tests, log gathering, and cleanup. If you'd like to discuss `kola` more feel free to reach out and I'd be more than happy to go through the project, current & future plans or any questions you might have.

We are initially planning to be running our automated testing against nightlies at a minimum but working our way towards per commit testing. `kola` produces an output folder (`_kola_temp`) which contains the logs for a test run (broken down per test and further per individual machine containing the serial console, journalctl output, and userdata passed to the instance). Providing public access to the artifacts should be as simple as archiving a tar of the contents & exposing the console log of whatever job is running the tests. [The `jenkins-os` repository contains Jenkins job we run as part of Container Linux](https://github.com/coreos/jenkins-os/tree/master/os/kola) if you want to look at an example of how this could be done (unfortunately the jenkins instance itself for Container Linux is not public though).

We don't currently have a concrete test plan, we're planning on porting / migrating some of the existing tests that make sense in the context of Fedora CoreOS but until we have some more decisions layed out about what will and will not be contained in the base OS that is still nebulous. Our general idea for what categories we're wanting to focus on in our testing is: Container Usage (docker/podman), Ignition, Upgrade, Metadata, K8s related (etcd, kubelet, etc.), SELinux, and a more general set (similar to the [`coreos.basic` & `coreos.internet` tests](https://github.com/coreos/mantle/blob/master/kola/tests/coretest/core.go))

Almost all of the test matrix listed under the [Atomic section](https://fedoraproject.org/wiki/Test_Results:Fedora_28_RC_1.1_Cloud#Atomic) are already present in some form in `kola` (for instance, the `base services start` test is performed against every machine spawned by the framework), except for `base service manipulation`.

[Here is a link](https://github.com/coreos/mantle#platform-credentials) that lists out the different platforms currently supported by the project (we're hoping to be working on adding OpenStack soon & finalizing our Azure support).

Once the initial fedora artifacts are being produced test days are indeed something that we'd be interested in having set up. Having additional users on the distribution hitting different configurations / scenarios than we might have considered is always helpful.

**Comment by ajeddeloh** [comment]
One thing that would be nice that we don't have today for CL is some sort of way to view kola logs in a browser without having to download the tarball.

**Comment by kparal** [comment]
Thanks @arithx for great info. It sounds like you push for all tests being automated. Is there anything you expect to be tested manually in the foreseeable future (e.g. at least with release candidates)? Can the kola test suite be executed against all supported cloud platforms, and locally (using qemu, I assume)?

In Fedora, there's heavy emphasis on systems being public and replicable (and of course running open source software). Exposing the test artifacts in public is definitely important, but would it be possible to make the Jenkins instance publicly accessible (read only, of course)?

Related to this, there's a [CI](https://fedoraproject.org/wiki/CI) initiative in Fedora (with Red Hat heavily investing into it), and it contains a [CI Pipeline](https://fedoraproject.org/wiki/CI/Pipeline) tool that runs inside Centos CI on Jenkins instances and executes tests important for Fedora (currently, test suites for freshly built packages or their pull requests are executed there). Have you heard of this/looked at this? There's a push to make it the universal driver for all testing needs in Fedora. I could put you in contact with the right people, if this looks like something you could use to execute your test suites publicly. Alternatively, a generic Jenkins instance from Centos CI could be used, I believe (without integrating with the CI Pipeline) in order to have public test machines.

Regarding nightlies, are there any experimental outputs being created already, so that I can have a look at them already (ideally with some short docs how to perform basic stuff :blush: )? Or is there some ETA for them?

Once we deal with the basics, I'd like to inform your team about the available QA processes that we have, for example how to propose a test day or how to ask for a freeze exception (if something is blocking your deliverable and we're currently in a freeze period) or how to document an uncomfortable bug that hasn't been solved yet (a "common bug"). Where's the best place to send this info - into this ticket, into the mailing list, into coreos discourse?

Thanks for your time.

**Comment by arithx** [comment]
@kparal, yes, our ultimate goal is for everything reasonable to be automated. Currently there isn't anything I'm expecting will be manually tested in the forseeable future. `kola` supports the following platforms currently (with a PR for adding Azure support): AWS, DigitalOcean, ESXi, GCE, Packet, & QEMU.

In Container Linux we have a similar approach of an easily replicable build pipeline & testing infrastructure. The [jenkins-os](https://github.com/coreos/jenkins-os) project contains our building scripts and resource definitions for overwriting specific credentials / release buckets. This is definitely something I personally want to continue doing moving forward on FCOS, as well as having a publicly accessible read-only jenkins instance.

On the Fedora STI front I have looked into it. The `kola` test framework doesn't quite fit into the standard mold dictated by the STI as I understand it, as it handles some of the responsibilities of both the testing system & test suite. I think that it would be more appropriate to have a generic Jenkins instance with credentials for different cloud environments loaded in it that we use to run the machines. @dustymabe might be able to speak more on if there are any current plans for where this infrastructure is planned to be ran from.

We currently have not published any experimental outputs, @dustymabe might be able to speak more on a potential ETA.

Of the places mentioned for posting the information I personally would lean towards the mailing list. It doesn't directly fit into the issue tracker and discourse in my opinion also doesn't fit the correct user group / use case for that information.

**Comment by kparal** [comment]
Thanks. A couple more questions, if you don't mind.

Do you have your own QA people in Fedora CoreOS team, or are you developers only?

Will the nightlies be built in Koji, or somewhere else? I assume the hosting and downloading process will be the same as for other Fedora variants, i.e. distributing them through Fedora mirrors system, right?

Regarding test suite results, we're trying to place all test results into a single place, that place being [ResultsDB](https://taskotron.fedoraproject.org/resultsdb/results) ([project](https://pagure.io/taskotron/resultsdb)). It would be great if your results could be submitted there as well, especially for latest Branched composes. We'd could keep an eye on the results, and inspect them for high profile nightly composes or release candidates. The result entry is a short simple json object that you send through a REST API (or a python library), and it contains a PASSED/FAILED outcome together with a link to test artifacts (logs etc, in your case probably a tarball or the jenkins results page, once it is public). I'd be happy to help you with that. (I guess that might need to wait until there are some FCOS images produced so that you have something to run `kola` suite on, right?)

And finally, is there something we can help you with from our side? That means anything related to Fedora or QA side of things.

**Comment by kparal** [comment]
> I'd love for someone to be involved in the Fedora CoreOS working group (roshi used to be our point man back in the day) and help us identify opportunities for improved testing and better collaboration with QA and the community. I believe coremodule has been interested in this role recently, but not sure if he still feels that way or not.

Yes, that would be me or coremodule, depending on the time of the day (I'm UTC+2, he's UTC-6, summer time included). How do we start? :)

**Comment by dustymabe** [comment]
> Yes, that would be me or coremodule, depending on the time of the day (I'm UTC+2, he's UTC-6, summer time included). How do we start? :)

basically join [communication channels](https://github.com/coreos/fedora-coreos-tracker#communication-channels-for-fedora-coreos) and/or [meetings](https://apps.fedoraproject.org/calendar/CoreOS) and be vocal!

**Comment by arithx** [comment]
@kparal for the ResultsDB are results entered per test case or per overarching test run. If it's per test case we'll have to write something that will extract that from the test run data (that should be simple enough if we need to, we produce a JSON result output for the entire test run).

**Comment by kparal** [comment]
> basically join communication channels and/or meetings and be vocal!

OK, I joined the channels. I'll be vocal when I understand what's going on (cloud is not my forte) and have something helpful to add :-)

> @kparal for the ResultsDB are results entered per test case or per overarching test run. If it's per test case we'll have to write something that will extract that from the test run data (that should be simple enough if we need to, we produce a JSON result output for the entire test run).

Both ways are possible. The defined test case can be specified in a dot notation to indicate subresults. In this case, you could have a single PASS/FAIL result for e.g. `coreos.kola` testcase (the `item` tested would be probably a compose ID or a specific image filename). Or you could have the overall result `coreos.kola`, but then also add `coreos.kola.sanity.boot` and `coreos.kola.deploy.aws` with individual subresults. It depends what the use case is for those results. If some other tool could make use of certain subresults (e.g. Bodhi or something), then it makes sense to have these results fine-grained in ResultsDB. For the beginning, until we have more uses for it (other than QA watching/getting notified of those), the single overall one would be, I think, completely satisfactory.

A hiccup I found out today is that we don't have any authentication support in ResultsDB. For the moment, we simply specify a set of IP addresses which can submit results. So, in order to show FCOS test results in ResultsDB, we would need one of these:
a) run `kola` inside Fedora Infrastructure
b) write a script that would run inside Fedora Infrastructure, listen for `kola` results on a remote server (using Jenkins API, or something custom) and transfer those results to ResultsDB
c) use the same solution as CI Pipeline does, which runs in Centos CI infra. They have their own ResultsDB instance that they submit results to, and we run a synchronization script that pulls their new results into our instance.
d) implement authentication in ResultsDB

Thoughts? :)

**Comment by dustymabe** [comment]
> A hiccup I found out today is that we don't have any authentication support in ResultsDB. For the moment, we simply specify a set of IP addresses which can submit results.

I think an easy target environment for us to run tests in would be centos CI (they have jenkins and openshift instances) and I think the problem of getting CentOS CI test results into ResultsDB has already been solved. Let me know if I'm wrong :)

**Comment by kparal** [comment]
> I think an easy target environment for us to run tests in would be centos CI (they have jenkins and openshift instances) and I think the problem of getting CentOS CI test results into ResultsDB has already been solved. Let me know if I'm wrong :)

Yes, that's point c). If you could use the same ResultsDB instance they already use inside CentosCI, then that would solve the issue of moving the results into the master Fedora instance. The synchronization process is already set up and working.

**Comment by glevand** [comment]
The [jenkins-os](https://github.com/coreos/jenkins-os) project was designed so that it can be run on any Jenkins system.  I have my own local Jenkins system here that uses jenkins-os to do the complete Container Linux CI (SDK build + OS build + kola tests + release).  I use this to do a complete check before submitting any significant  code change, and also when fixing bugs or adding enhancements to jenkins-os itself.  I think this feature is of huge value and would like to see brought forward to FCOS.

**Comment by kparal** [comment]
I wonder, who's going to be at Flock next week? It would be great if I could meet you in person. I see that Dusty and Benjamin have a talk about CoreOS. I'll stop by and say hello. Anyone else I can look out for at the conference?

**Comment by cgwalters** [comment]
>  I think this feature is of huge value and would like to see brought forward to FCOS.

I'm in 100% agreement with that, although having used Jenkins a lot - the thing is Jenkins wants to be a scheduler + owner of compute resources, and Kubernetes is simply better at that.   Some prototyping we're doing for Red Hat CoreOS today uses Jenkins but I am pushing forward with preparing for a world where the jobs execute from a container that can be scheduled as a pod.  Jenkins would be a toplevel orchestrator there.  What do you think of that?

**Comment by jlebon** [comment]
Was chatting with @AdamWill today about how we'd go about adding FCOS tests to Bodhi updates. He provided a bunch of links to implementations of how this all works today. Pasting (abbreviated) convo for reference:

```
< jlebon> btw, one thing related to Fedora testing and FCOS we've discussed a few times is having FCOS tests
          in bodhi for the critical packages like systemd, kernel, selinux stuff, etc...
< adamw> you'd presumably just want to build a new image when a relevant update appears, right?
< adamw> then run that through whatever test systems
< adamw> i do know exactly how to do it, yeah
< jlebon> adamw: exactly yes. it's more all the glue/APIs in between that we'll need help with i think
< jlebon> (and of course, actually socializing this more broadly before doing it)
< adamw> so the way openqa does it is, it listens for two bodhi messages:
< adamw> bodhi.update.edit
< adamw> and
< adamw> bodhi.update.request.testing
< adamw> you can read lots of useful info about the update out of the message, plus you can query bodhi API
         if you need more
< jlebon> gotcha. and how do you report back to bodhi?
< adamw> for openQA i have it test all critical path updates (that info is in the message) plus all updates
         containing any package on a list of non-critpath-but-interesting-to-test packages
< adamw> that goes via resultsdb and greenwave
< adamw> let me look it all up again to make sure i'm not lying to you :D
< adamw> yeah that's it, right. so. see here:
< adamw> https://github.com/fedora-infra/bodhi/blob/develop/bodhi/server/models.py#L2094
< adamw> this is kind of a weird and indirect way to do it, but it's how it works...bodhi queries greenwave
         for a "decision" on the update, i.e., whether it passes or fails some gating criteria. but it asks
         for a "verbose" decision, which means greenwave *also* includes all the resultsdb results it found
         based on the query params
< adamw> bodhi then displays those results in the UI
< adamw> so, you see there that it uses two query subjects? basically it's telling greenwave "look up
         results with type: 'koji_build' and item: [build] for all builds in this update, and also results
         with type: 'bodhi_update' and item: [alias of this update]"
< adamw> and it'll display all the results it gets back from that query in the web UI
< adamw> so basically: you need to report your test results to resultsdb in a form that query will find
< jlebon> do you have the openQA code which submits results to resultsdb handy?
< adamw> if you're actually testing at the update level, you'll probably want to report at that level too
< adamw> yup
< adamw> https://pagure.io/fedora-qa/fedora_openqa/blob/master/f/src/fedora_openqa/report.py#_344
< adamw> it uses a thing i wrote called resultsdb_conventions which was an attempt to kinda provide a
         standardized way to report resultsdb results
< adamw> with the idea being we could then rely on results all following the same format
< adamw> it didn't really get adopted by anyone, though. le sigh
< adamw> https://pagure.io/taskotron/resultsdb_conventions
< adamw> but you could use it! :) then we would be confident openqa and kola results would be in a similar
         form.
< adamw> but to satisfy bodhi, you'd really just need to make sure the result had item: (alias of update) and
         type: 'bodhi_update'.
< adamw> if i was redoing it today i'd base it around json schema like the ci-messages thing, but ah well.
< jlebon> when you say "alias of update" you're talking about the advisory ID, right? e.g. FEDORA-2020-... ?
< adamw> yeah
< adamw> bodhi calls it the 'alias' internally
< adamw> (and in api docs and stuff)
< adamw> for reference, an openQA result looks like
         https://taskotron.fedoraproject.org/resultsdb/results/39427194
< adamw> the koji_build query subject is basically to capture results from fedora CI, which tests at the
         package level
< adamw> it doesn't (yet) have the ability to test update package sets together, afaik
```

**Comment by cgwalters** [comment]
I think this would be good, but we really need to also test the full ostree commit before it ships to humans.  Doing that is like 90% of the whole value of an image-based/ostree approach.  And we are doing that today with FCOS.

Also related to this, there's an immense amount of value in what Fedora CoreOS is doing today with staged rollouts...and really all of this gets into rebasing Silverblue on FCOS, aka https://github.com/cgwalters/fedora-silverblue-config

**Comment by jlebon** [comment]
Related: [Plan / proposal: enable openQA update testing and potentially gating on Rawhide updates](https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/message/JK4TZIRE3QP32VIT5BOXTIPMKRATTAAM/)

**Comment by miabbott** [comment]
I was talking with @jlebon today and we discussed the idea of using ostree native containers [[1](https://coreos.github.io/rpm-ostree/container/)][[2](https://www.fedoraproject.org/wiki/Changes/OstreeNativeContainerStable)] as a way to run tests on updates submitted to Bodhi.

It should be possible to take a Fedora CoreOS base image (i.e. quay.io/fedora/fedora-coreos:testing) and layer the updated RPM into a new container image. That resulting container image could then be used as a target for running a set of kola tests that could provide feedback on whether the Bodhi update was good or bad. (In some cases, a full compose may be required for more rigorous tests)

We could introduce this kind of testing as a non-gating tests on Bodhi updates for the packages that are relevant to Fedora CoreOS with the goal of making them gating on packages deemed critical for FCOS (or all of Fedora).

This would require a lot of education, outreach, and documentation so that packagers could appropriately respond to failing test results and not overload the CoreOS team with debug requests.

**Comment by AdamWill** [comment]
We do something along those lines for Silverblue in openQA already, except we don't use layering; we just flat out build a new ostree with the test package in it, a new ostree installer image with the new ostree image, run an install, and test the installed system. Layering would definitely be faster!

If we get to https://fedoraproject.org/wiki/Changes/OstreeNativeContainerStable for Silverblue etc., hopefully I'd be able to use layering for that test too and make it much faster...

**Comment by cgwalters** [comment]
> If we get to https://fedoraproject.org/wiki/Changes/OstreeNativeContainerStable for Silverblue etc., 

It already exists, https://quay.io/repository/fedora/fedora-silverblue

**Comment by AdamWill** [comment]
Yeah, I know it exists, but I mean, if it becomes the default/official Silverblue deployment. openQA has to focus on testing the thing we're currently shipping officially.

**Comment by jlebon** [comment]
Some development on this in https://github.com/coreos/coreos-ci/pull/49.
